botorch_community/acquisition/bll_thompson_sampling.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from __future__ import annotations
+
+import numpy as np
+import scipy
+
+import torch
+
+from botorch.logging import logger
+
+from botorch_community.models.blls import AbstractBLLModel
+from torch.func import grad
+
+
+class BLLMaxPosteriorSampling:
+    def __init__(
+        self,
+        model: AbstractBLLModel,
+        num_restarts: int = 10,
+        bounds: torch.Tensor | None = None,
+        discrete_inputs: bool = False,
+    ):
+        """
+        Implements Maximum Posterior Sampling for Bayesian Linear Last (VBLL) models.
+
+        This class provides functionality to sample from the posterior distribution of a
+        BLL model, with optional optimization to refine the sampling process.
+
+        Args:
+            model: The VBLL model from which posterior samples are drawn. Must be an
+                instance of `AbstractBLLModel`.
+            num_restarts: Number of restarts for optimization-based sampling.
+                Defaults to 10.
+            bounds: Tensor of shape (2, num_inputs) specifying the lower and upper
+                bounds for sampling. If None, defaults to [(0, 1)] for each input
+                dimension.
+            discrete_inputs: If True, assumes the input space is discrete and will be
+                provided in __call__. Defaults to False.
+
+        Raises:
+            ValueError:
+                If the provided `model` is not an instance of `AbstractBLLModel`.
+
+        Notes:
+            - If `bounds` is not provided, the default range [0,1] is assumed for each
+            input dimension.
+        """
+        if not isinstance(model, AbstractBLLModel):
+            raise ValueError(
+                f"Model must be an instance of AbstractBLLModel, is {type(model)}"
+            )
+
+        self.model = model
+        self.device = model.device
+        self.discrete_inputs = discrete_inputs
+        self.num_restarts = num_restarts
+
+        if bounds is None:
+            # Default bounds [0,1] for each input dimension
+            self.bounds = [(0, 1)] * self.model.num_inputs
+            self.lb = torch.zeros(
+                self.model.num_inputs, dtype=torch.float64, device=torch.device("cpu")
+            )
+            self.ub = torch.ones(
+                self.model.num_inputs, dtype=torch.float64, device=torch.device("cpu")
+            )
+        else:
+            # Ensure bounds are on CPU for compatibility with scipy.optimize.minimize
+            self.lb = bounds[0, :].cpu()
+            self.ub = bounds[1, :].cpu()
+            self.bounds = [tuple(bound) for bound in bounds.T.cpu().tolist()]
+
+    def __call__(
+        self, X_cand: torch.Tensor = None, num_samples: int = 1
+    ) -> torch.Tensor:
```

Comment:
Can you add a docstring describing this method, including the expected input and return shapes of the tensors?